src/server/mod.rs

```diff
@@ -270,49 +269,25 @@ impl<State: Send + Sync + 'static> Server<State> {
     /// Middleware can only be added at the "top level" of an application,
     /// and is processed in the order in which it is applied.
     pub fn middleware(&mut self, m: impl Middleware<State>) -> &mut Self {
-        self.middleware.push(Arc::new(m));
+        let middleware = Arc::get_mut(&mut self.middleware).unwrap();
```

Comment:
Uhm why are these `Arc::get_mut` `unwrap`s safe? Pretty sure you are not guranteed to be the only holder here.

Reply:
> Pretty sure you are not guranteed to be the only holder here.

Notice how we're not cloning the Arc, we're only giving a reference to the inner value. This is the same semantics we currently have. The `Arc` is only cloned once we start listening for requests, after which `self` is taken by value and becomes unavailable.

Maybe we can turn this into an `expect` so in the off-chance someone tries something bizarre it fails with a better error. But this is effectively the same semantics we had before.

Reply:
Okay, added error messages for the failures. I think this should do it!

Reply:
Oh also for reference: this is how we used to go between `Server` and `Service` before:

```rust
    pub fn into_http_service(self) -> Service<State> {
        Service {
            router: Arc::new(self.router),
            state: Arc::new(self.state),
            middleware: Arc::new(self.middleware),
        }
    }
```

All it does is put the extra fields in an `Arc` and make certain methods available. Those methods currently still take `self` / `&mut self` which means we're protected by the borrow checker. The main difference is that we now start by putting the values in an `Arc` in the first place instead of moving them into one later. This is semantically the same.

Reply:
I see, thank you